### DISCUSSION TITLE
Inferring correct overload with kwargs and default arg

### DISCUSSION THREAD

**Discussion by davetapley** [microsoft/pyright#6995] [General]
I'm not sure if this is _bug_, an unsupported edge case, or me making a mistake 😁 

`pandas` [`to_csv`](https://github.com/pandas-dev/pandas-stubs/blob/56eafc1ee579bfc3a145c71ec5deb0d5dd7b00ba/pandas-stubs/core/generic.pyi#L238-L286C11) has two overloads:
1. If `path_or_buf` is `None` it returns `str`
2. Otherwise it returns `None`.

It takes quite a lot of kwargs, so for readability I want to build them up in a `dict` first and splat them in with `**`, like this:

```py
opts: dict[str, Any] = {'index': False, 'header': False, 'sep': self.config.delimiter}
        if self.config.date_format:
            opts['date_format'] = self.config.date_format
        # ...
        out = df.to_csv(**opts)
```

That yields
```
Type of "out" is unknownPylance[reportUnknownVariableType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnknownVariableType)
```

But if I explicitly give `path_or_buf` then it correctly infers `str`:

```py
        out = df.to_csv(path_or_buf=None, **opts)
```

Thoughts?

---

Additionally: my `[dict](opts: dict[str, Any])` could be narrowed with [PEP 692](https://peps.python.org/pep-0692/), so perhaps when that's in then it could figure it out?

**Comment by twoertwein** [comment]
I don't fully understand your use case, but you can definitely use TypedDict and pass all arguments in one typed dict:

```py
from typing import TypedDict

import pandas as pd

df = pd.DataFrame()

class ReadCSVNone(TypedDict):
    path_or_buf: None
    sep: str
    # and all the other arguments
    
kwargs: ReadCSVNone = {"sep": ",", "path_or_buf": None}

reveal_type(df.to_csv(**kwargs)) # str
```

**Reply by erictraut** [reply]
It's difficult to fully diagnose the issue because the code you've posted isn't self-contained. It refers to symbols and types that are not defined.

It looks like you've enabled the `reportUnknownVariableType` diagnostic check. This isn't enabled by default unless you've enabled "strict" mode. Was it your intent to enable this? If you enable this, pyright will tell you when an implicit `Any` is assigned to a variable. In this case, it looks like pyright couldn't unambiguously determine which overload to use because you have effectively blinded it by passing all of the keyword arguments through a `dict[str, Any]`. If you use a `TypedDict` as @twoertwein suggests, that should eliminate the ambiguity.
